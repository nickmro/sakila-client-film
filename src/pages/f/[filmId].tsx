import React, { useMemo } from 'react';
import { NextPage } from 'next';
import styled from 'styled-components';
import { getFilm } from '../../apollo/film';
import HeroImage from '../../components/HeroImage';
import { filmDuration } from '../../helpers/film';
import { Film } from '../../types/film';

const DetailsContainer = styled.div`
  width: 100%;
`;

const Details = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  padding: 25px;
`;

const Title = styled.h1`
  font-size: 28px;
  margin-top: 0;
  margin-bottom: 10px;
`;

const Tags = styled.div`
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  margin-bottom: 10px;
`;

const Tag = styled.span`
  display: flex;

  &:not(:first-child) {
    margin: 0 0 0 15px;
  }
`;

const Actors = styled.div`
  font-size: 14px;
  margin-bottom: 10px;
`;

const Description = styled.div`
  font-size: 18px;
  margin-bottom: 10px;
`;

export interface FilmPageProps {
  film?: Film;
}

export const FilmPage: NextPage<FilmPageProps> = (props) => {
  const { film } = props;

  const actors = useMemo<string | undefined>(() => film?.actors
    .map((actor) => `${actor.firstName} ${actor.lastName}`)
    .join(', '), [film]);

  return (
    <>
      { !film ? (
        <p>Not Found</p>
      ) : (
        <>
          <HeroImage />
          <DetailsContainer>
            <Details>
              <Title className="title">{film.title}</Title>
              <Tags>
                <Tag className="release-year">{film.releaseYear}</Tag>
                <Tag className="rating">{film.rating}</Tag>
                <Tag className="duration">{filmDuration(film.length)}</Tag>
              </Tags>
              <Actors>
                <span className="actors">{actors}</span>
              </Actors>
              <Description>
                <span className="description">{film.description}</span>
              </Description>
            </Details>
          </DetailsContainer>
        </>
      )}
    </>
  );
};

FilmPage.getInitialProps = async (ctx) => {
  const { filmId: filmIdParam } = ctx.query;

  try {
    const filmId = filmIdParam as string;

    const { data } = await getFilm(parseInt(filmId, 10));
    return { film: data?.film };
  } catch {
    return { film: undefined };
  }
};

export default FilmPage;
