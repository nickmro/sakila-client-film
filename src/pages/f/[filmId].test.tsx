import React from 'react';
import renderer from 'react-test-renderer';
import { FilmPage } from './[filmId]';

const film = {
  filmId: 1,
  title: 'ACADEMY DINOSAUR',
  description: 'A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies',
  length: 86,
  rating: 'PG',
  releaseYear: 2006,
  actors: [
    {
      actorId: 1,
      firstName: 'PENELOPE',
      lastName: 'GUINESS',
    },
    {
      actorId: 2,
      firstName: 'CHRISTIAN',
      lastName: 'GABLE',
    },
  ],
};

describe('FilmPage', () => {
  beforeEach(() => {
    jest.mock('../../apollo/film');
  });

  test('it renders', () => {
    const filmPage = renderer.create(<FilmPage film={film} />);
    expect(filmPage).toBeTruthy();
  });

  test('it renders the film title', () => {
    const filmPage = renderer.create(<FilmPage film={film} />);
    const title = filmPage.root.findByProps({ className: 'title' }).props.children;
    expect(title).toEqual(film.title);
  });

  test('it renders the description', () => {
    const filmPage = renderer.create(<FilmPage film={film} />);
    const description = filmPage.root.findByProps({ className: 'description' }).props.children;
    expect(description).toEqual(film.description);
  });

  test('it renders the rating', () => {
    const filmPage = renderer.create(<FilmPage film={film} />);
    const rating = filmPage.root.findByProps({ className: 'rating' }).props.children;
    expect(rating).toEqual(film.rating);
  });

  test('it renders the duration', () => {
    const filmPage = renderer.create(<FilmPage film={film} />);
    const duration = filmPage.root.findByProps({ className: 'duration' }).props.children;
    expect(duration).toEqual('1h 26m');
  });

  test('it renders the release year', () => {
    const filmPage = renderer.create(<FilmPage film={film} />);
    const releaseYear = filmPage.root.findByProps({ className: 'release-year' }).props.children;
    expect(releaseYear).toEqual(film.releaseYear);
  });

  test('it renders the actors', () => {
    const filmPage = renderer.create(<FilmPage film={film} />);
    const actors = filmPage.root.findByProps({ className: 'actors' }).props.children;
    expect(actors).toEqual('PENELOPE GUINESS, CHRISTIAN GABLE');
  });
});
