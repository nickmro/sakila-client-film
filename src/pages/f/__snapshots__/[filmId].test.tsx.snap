// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmPage it renders 1`] = `
Array [
  <div
    className="HeroImage__Container-ljnxwg-0 PIYbc"
  >
    <picture>
      <div
        className="HeroImage__Image-ljnxwg-1 iIQYpF"
      />
    </picture>
  </div>,
  <div
    className="filmId__DetailsContainer-q2shma-0 euPiVg"
  >
    <div
      className="filmId__Details-q2shma-1 pGFZW"
    >
      <h1
        className="filmId__Title-q2shma-2 jwHjfV title"
      >
        ACADEMY DINOSAUR
      </h1>
      <div
        className="filmId__Tags-q2shma-3 kCyOOj"
      >
        <span
          className="filmId__Tag-q2shma-4 bzxaUC release-year"
        >
          2006
        </span>
        <span
          className="filmId__Tag-q2shma-4 bzxaUC rating"
        >
          PG
        </span>
        <span
          className="filmId__Tag-q2shma-4 bzxaUC duration"
        >
          1h 26m
        </span>
      </div>
      <div
        className="filmId__Actors-q2shma-5 kMvRcI"
      >
        <span
          className="actors"
        >
          PENELOPE GUINESS, CHRISTIAN GABLE
        </span>
      </div>
      <div
        className="filmId__Description-q2shma-6 hHbVjN"
      >
        <span
          className="description"
        >
          A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies
        </span>
      </div>
    </div>
  </div>,
]
`;
